proxy_test.go

```diff
@@ -596,3 +604,57 @@ func TestShimSignal(t *testing.T) {
 
 	rig.Stop()
 }
+
+// TestShimConnectAfterExeccmd tests we correctly wait for the shim to connect
+// before forwarding the execcmd to hyperstart.
+func TestShimConnectAfterExeccmd(t *testing.T) {
+	rig := newTestRig(t)
+	rig.Start()
+
+	// Register new VM, asking for tokens. We use the assumption the same
+	// connection can be used for ConnectShim, which is true in the tests.:62
+	ctlSocketPath, ioSocketPath := rig.Hyperstart.GetSocketPaths()
+	ret, err := rig.Client.RegisterVM(testContainerID, ctlSocketPath, ioSocketPath,
+		&goapi.RegisterVMOptions{NumIOStreams: 1})
+	assert.Nil(t, err)
+	assert.Equal(t, 1, len(ret.IO.Tokens))
+	tokens := ret.IO.Tokens
+
+	// Send an execcmd. Since we don't have the corresponding shim yet, this
+	// should time out.
+	oldTimeout := waitForShimTimeout
+	waitForShimTimeout = 20 * time.Millisecond
+	execcmd := hyperstart.ExecCommand{
+		Container: testContainerID,
+		Process: hyperstart.Process{
+			Args: []string{"/bin/sh"},
+		},
+	}
+	err = rig.Client.HyperWithTokens("execcmd", tokens, &execcmd)
+	assert.NotNil(t, err)
+	assert.True(t, strings.Contains(err.Error(), "timeout"))
+	waitForShimTimeout = oldTimeout
+
+	// Send an execcmd again, but this time will connect the shim just after the command.
+	shimConn := rig.ServeNewClient()
+	shim := newShimRig(t, shimConn, tokens[0])
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		time.Sleep(20 * time.Millisecond)
```

Comment:
Second occurence of `20 * time.Millisecond` - might be worth a variable for this?